influxql/call_iterator.go

```diff
@@ -500,6 +500,163 @@ func IntegerMedianReduceSlice(a []IntegerPoint) []FloatPoint {
+		createFn := func() (StringPointAggregator, StringPointEmitter) {
+			fn := NewStringSliceFuncReducer(StringModeReduceSlice)
+			return fn, fn
+		}
+		return &stringReduceStringIterator{input: newBufStringIterator(input), opt: opt, create: createFn}, nil
+	case BooleanIterator:
+		createFn := func() (BooleanPointAggregator, BooleanPointEmitter) {
+			fn := NewBooleanSliceFuncReducer(BooleanModeReduceSlice)
+			return fn, fn
+		}
+		return &booleanReduceBooleanIterator{input: newBufBooleanIterator(input), opt: opt, create: createFn}, nil
+
+	default:
+		return nil, fmt.Errorf("unsupported median iterator type: %T", input)
+	}
+}
+
+// FloatModeReduceSlice returns the mode value within a window.
+func FloatModeReduceSlice(a []FloatPoint) []FloatPoint {
+	if len(a) == 1 {
+		return a
+	}
+
+	// fmt.Println(a[0])
+	sort.Sort(floatPointsByValue(a))
+
+	mostFreq := 0
+	currFreq := 0
+	currMode := a[0].Value
+	mostMode := a[0].Value
+	mostTime := a[0].Time
+
+	for _, p := range a {
+		if p.Value != currMode {
+			currFreq = 1
+			currMode = p.Value
+			continue
+		}
+		currFreq++
+		if mostFreq > currFreq || (mostFreq == currFreq && p.Time > mostTime) {
+			continue
+		}
+		mostFreq = currFreq
+		mostMode = p.Value
+		mostTime = p.Time
+	}
+
+	return []FloatPoint{{Time: ZeroTime, Value: mostMode}}
+}
+
+// IntegerModeReduceSlice returns the mode value within a window.
+func IntegerModeReduceSlice(a []IntegerPoint) []IntegerPoint {
+	if len(a) == 1 {
+		return a
+	}
+	sort.Sort(integerPointsByValue(a))
+
+	mostFreq := 0
+	currFreq := 0
+	currMode := a[0].Value
+	mostMode := a[0].Value
+	mostTime := a[0].Time
+
+	for _, p := range a {
+		if p.Value != currMode {
+			currFreq = 1
+			currMode = p.Value
+			continue
+		}
+		currFreq++
+		if mostFreq > currFreq || (mostFreq == currFreq && p.Time > mostTime) {
+			continue
+		}
+		mostFreq = currFreq
+		mostMode = p.Value
+		mostTime = p.Time
+	}
+
+	return []IntegerPoint{{Time: ZeroTime, Value: mostMode}}
+}
+
+// StringModeReduceSlice returns the mode value within a window.
+func StringModeReduceSlice(a []StringPoint) []StringPoint {
+	if len(a) == 1 {
+		return a
+	}
+
+	sort.Sort(stringPointsByValue(a))
+
+	mostFreq := 0
+	currFreq := 0
+	currMode := a[0].Value
+	mostMode := a[0].Value
+	mostTime := a[0].Time
+
+	for _, p := range a {
+		if p.Value != currMode {
+			currFreq = 1
+			currMode = p.Value
+			continue
+		}
+		currFreq++
+		if mostFreq > currFreq || (mostFreq == currFreq && p.Time > mostTime) {
+			continue
+		}
+		mostFreq = currFreq
+		mostMode = p.Value
+		mostTime = p.Time
+	}
+
+	return []StringPoint{{Time: ZeroTime, Value: mostMode}}
+}
+
+// BooleanModeReduceSlice returns the mode value within a window.
+func BooleanModeReduceSlice(a []BooleanPoint) []BooleanPoint {
+	if len(a) == 1 {
+		return a
+	}
+
+	trueFreq := 0
+	falsFreq := 0
+	mostMode := false
+
+	for _, p := range a {
+		if p.Value {
+			trueFreq++
+			continue
+		}
+		// a nil value is treated as false
+		falsFreq++
```

Comment:
But wouldn't values passed to iterator will be either `true` or `false` . I made the comment assuming that it will be get all `nil` values as false.